### PR TITLE
ddl: fix a bug when drop a column with single index

### LINKED ISSUE #32079: index out of range in the schrddl
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

schrddl
```
 [mysql] 2022/01/31 16:43:30 packets.go:36: unexpected EOF
2022/01/31 16:43:30 util.go:55: [warning] Failed to disable TiKV GC
2022/01/31 16:43:30 dml_ops.go:78: [info] [dml] [instance 2] INSERT INTO `e0fc0bcc-df65-4cae-b44a-3ef1557af3a2` SET `3f224854-bb26-454a-bb50-5ee3eef722e6` = 'mma8,ut,py0e,qxu,6,n5,yr8v,1,36dv', `1762176d-fa14-4b52-95e4-b17e745b686d` = '2106296027', err: <nil>
2022/01/31 16:43:30 dml_ops.go:78: [info] [dml] [instance 0] UPDATE `116b54a2-2d96-415e-88b1-68757e6e4e33` SET `bb8dbd27-c8da-4712-8f47-8b6936857654` = '50', `08bdf016-4cfa-493a-b12d-8db9216b4e31` = 'ivr0kqz1gop' WHERE `bb8dbd27-c8da-4712-8f47-8b6936857654` = '-24' AND `08bdf016-4cfa-493a-b12d-8db9216b4e31` = 'p77lv8ufhjvjg3tqs5r', err: <nil>
2022/01/31 16:43:30 ddl.go:96: [fatal] [error] [instance 1] ERROR: Error 1105: runtime error: index out of range [3] with length 3
/build/schrddl/ddl/dml_ops.go:84: Error when executing SQL: UPDATE `849e9725-6a7a-4332-aa6a-06f983a316ad` SET `dfcd580a-2867-4304-8b06-b140d8626dee` = '1.2373593389856343', `10ef9fab-38b2-4943-9dcd-9a0027ff9310` = '0', `484e923c-66d4-4ba1-bd5c-3ef56ec04126` = 'gc,w,n3,8xom,u4,f,uy,osk6j,5'
======== DEBUG BEGIN  ========
Dumping expected contents for table `849e9725-6a7a-4332-aa6a-06f983a316ad`:
Comment: 742f48ae-f1bb-440b-b7e2-f59a3bb3953f
Charset: gbk, Collate: gbk_chinese_ci
ShardRowId: 2
AutoID: 0
## Non-Primary Indexes: 
Index #0: Name = `0d351f7b-5c9c-4e9f-9411-9ae892828a1b`, Columnns = [`5d168842-48d8-4b31-a3bc-acd2cb87871c`, ]
Index #1: Name = `3d4e5dbc-c152-48d0-a853-1ea79a92ed7c`, Columnns = [`484e923c-66d4-4ba1-bd5c-3ef56ec04126`, ]
## Columns: 
Column #0: Name = `484e923c-66d4-4ba1-bd5c-3ef56ec04126`, Definition = SET("gc", "w", "n3", "8xom", "u4", "f", "ou", "uy", "osk6j", "j", "5") NULL DEFAULT 'gc,w', isPrimaryKey = false, used in 1 indexes
Column #1: Name = `10ef9fab-38b2-4943-9dcd-9a0027ff9310`, Definition = BOOL NULL DEFAULT '1', isPrimaryKey = false, used in 0 indexes
Column #2 [DELETED]: Name = `5d168842-48d8-4b31-a3bc-acd2cb87871c`, Definition = DECIMAL(60,5) NULL DEFAULT '1583508031665035148785366080188772689918644934209064113.55271', isPrimaryKey = false, used in 0 indexes
Column #3: Name = `dfcd580a-2867-4304-8b06-b140d8626dee`, Definition = DOUBLE NULL DEFAULT '1.4607072349530976', isPrimaryKey = false, used in 0 indexes
## Values (number of rows = 1): 
#0   :          n3,       <nil>,       <nil>,       <nil>, 
======== DEBUG END ========
```

tidb 
```
 [2022/01/31 16:43:30.166 +00:00] [ERROR] [conn.go:1031] ["connection running loop panic"] [conn=17] [lastSQL="UPDATE `849e9725-6a7a-4332-aa6a-06f983a316ad` SET `dfcd580a-2867-4304-8b06-b140d8626dee` = '1.2373593389856343', `10ef9fab-38b2-4943-9dcd-9a0027ff9310` = '0', `484e923c-66d4-4ba1-bd5c-3ef56ec04126` = 'gc,w,n3,8xom,u4,f,uy,osk6j,5'"] [err="runtime error: index out of range [3] with length 3"] [stack="goroutine 1700 [running]:\ngithub.com/pingcap/tidb/server.(*clientConn).Run.func1()\n\t/tidb/server/conn.go:1029 +0x8f\npanic({0x39661e0, 0xc00332f230})\n\t/usr/local/go/src/runtime/panic.go:1038 +0x215\ngithub.com/pingcap/tidb/executor.(*ExecStmt).Exec.func1()\n\t/tidb/executor/adapter.go:362 +0x3c6\npanic({0x39661e0, 0xc00332f230})\n\t/usr/local/go/src/runtime/panic.go:1038 +0x215\ngithub.com/pingcap/tidb/table/tables.(*TableCommon).rebuildIndices(0xc00186c900, {0x418e8f8, 0xc00016d800}, {0x4159a30, 0xc00016d810}, {0x412f248, 0x5ea5898}, {0xc000578e4c, 0x3, 0x4}, ...)\n\t/tidb/table/tables/tables.go:468 +0x645\ngithub.com/pingcap/tidb/table/tables.(*TableCommon).UpdateRecord(0xc00186c900, {0x40e2010, 0xc001d36ba0}, {0x418e8f8, 0xc00016d800}, {0x412f248, 0x5ea5898}, {0xc00198afc0, 0x3, 0x4}, ...)\n\t/tidb/table/tables/tables.go:411 +0x1d89\ngithub.com/pingcap/tidb/executor.updateRecord({0x40e2010, 0xc001d36ba0}, {0x418e8f8, 0xc00016d800}, {0x412f248, 0x5ea5898}, {0xc00198afc0, 0x3, 0x4}, {0xc00198b0e0, ...}, ...)\n\t/tidb/executor/write.go:203 +0x1665\ngithub.com/pingcap/tidb/executor.(*UpdateExec).exec(0xc0021cfb00, {0x40e2010, 0xc001d36ba0}, 0x4, {0xc00198afc0, 0x1, 0x4}, {0xc00198b0e0, 0x4, 0x4})\n\t/tidb/executor/update.go:192 +0x3fc\ngithub.com/pingcap/tidb/executor.(*UpdateExec).updateRows(0xc0021cfb00, {0x40e2010, 0xc001d36ba0})\n\t/tidb/executor/update.go:312 +0x792\ngithub.com/pingcap/tidb/executor.(*UpdateExec).Next(0xc0021cfb00, {0x40e2010, 0xc001d36270}, 0x20)\n\t/tidb/executor/update.go:226 +0x9f\ngithub.com/pingcap/tidb/executor.Next({0x40e2010, 0xc001d36270}, {0x40ec138, 0xc0021cfb00}, 0xc0014cea00)\n\t/tidb/executor/executor.go:286 +0x4a2\ngithub.com/pingcap/tidb/executor.(*ExecStmt).handleNoDelayExecutor(0xc0023a2680, {0x40e2010, 0xc001d36270}, {0x40ec138, 0xc0021cfb00})\n\t/tidb/executor/adapter.go:635 +0x59f\ngithub.com/pingcap/tidb/executor.(*ExecStmt).handleNoDelay(0xc0023a2680, {0x40e2010, 0xc001d36270}, {0x40ec138, 0xc0021cfb00}, 0x0)\n\t/tidb/executor/adapter.go:490 +0x1f9\ngithub.com/pingcap/tidb/executor.(*ExecStmt).Exec(0xc0023a2680, {0x40e2010, 0xc001d36270})\n\t/tidb/executor/adapter.go:439 +0x92e\ngithub.com/pingcap/tidb/session.runStmt({0x40e2010, 0xc0034339e0}, 0xc00016d800, {0x40f3a68, 0xc0023a2680})\n\t/tidb/session/session.go:1880 +0x55d\ngithub.com/pingcap/tidb/session.(*session).ExecuteStmt(0xc00016d800, {0x40e2010, 0xc0034339e0}, {0x41179e0, 0xc0014abe40})\n\t/tidb/session/session.go:1757 +0xb25\ngithub.com/pingcap/tidb/server.(*TiDBContext).ExecuteStmt(0xc00167a570, {0x40e2010, 0xc0034339e0}, {0x41179e0, 0xc0014abe40})\n\t/tidb/server/driver_tidb.go:220 +0x44\ngithub.com/pingcap/tidb/server.(*clientConn).handleStmt(0xc00163ac80, {0x40e1f68, 0xc002238c00}, {0x41179e0, 0xc0014abe40}, {0x60fc200, 0x0, 0x0}, 0x1)\n\t/tidb/server/conn.go:1977 +0x167\ngithub.com/pingcap/tidb/server.(*clientConn).handleQuery(0xc00163ac80, {0x40e1f68, 0xc002238c00}, {0xc003522ff1, 0xe6})\n\t/tidb/server/conn.go:1846 +0x770\ngithub.com/pingcap/tidb/server.(*clientConn).dispatch(0xc00163ac80, {0x40e2010, 0xc001633950}, {0xc003522ff0, 0xe7, 0xe7})\n\t/tidb/server/conn.go:1341 +0xead\ngithub.com/pingcap/tidb/server.(*clientConn).Run(0xc00163ac80, {0x40e2010, 0xc001633950})\n\t/tidb/server/conn.go:1091 +0x253\ngithub.com/pingcap/tidb/server.(*Server).onConn(0xc000b74ea0, 0xc00163ac80)\n\t/tidb/server/server.go:554 +0xfa6\ncreated by github.com/pingcap/tidb/server.(*Server).startNetworkListener\n\t/tidb/server/server.go:451 +0x5cf\n"] 
```

### PR DESCRIPTION
**Pull request by wjhuang2016**
Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

### What problem does this PR solve?

Issue Number: close #32079

Problem Summary:
The problem is that, there is a state that column is deleteOnly and the index is deleOnly. To delete an index, we need to get the column data. But since the column is deleteOnly, we can't get it's value.
### What is changed and how it works?
Remove the index from the schema before the column enters the deleteOnly state.
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: ddl/column.go [hunk 1]
```diff
@@ -598,7 +606,7 @@ func onDropColumn(t *meta.Meta, job *model.Job) (ver int64, _ error) {
 	case model.StateDeleteOnly:
 		// delete only -> reorganization
 		colInfo.State = model.StateDeleteReorganization
-		setIndicesState(idxInfos, model.StateDeleteReorganization)
+		//setIndicesState(idxInfos, model.StateDeleteReorganization)
```
Comment (mjonss) [inline ddl/column.go:609]:
```suggestion
```

File: ddl/db_test.go [hunk 2]
```diff
@@ -7772,3 +7772,57 @@ func (s *testSerialDBSuite) TestAddGeneratedColumnAndInsert(c *C) {
...
+				return
+			}
+		case model.StateWriteReorganization:
+			if job.SchemaState == model.StateWriteReorganization && times == 0 {
+				_, checkErr = tk1.Exec("update t set id = 16 where id=160;")
+				if checkErr != nil {
+					return
+				}
+				currJob = job
+				times++
```
Comment (mjonss) [inline ddl/db_test.go:7817]:
```suggestion
```

File: ddl/db_test.go [hunk 3]
```diff
@@ -7772,3 +7772,57 @@ func (s *testSerialDBSuite) TestAddGeneratedColumnAndInsert(c *C) {
...
+	var checkErr error
+	tk1 := testkit.NewTestKit(c, s.store)
+	_, checkErr = tk1.Exec("use Issue32079")
+
+	d := s.dom.DDL()
+	hook := &ddl.TestDDLCallback{Do: s.dom}
+	var currJob *model.Job
+	ctx := mock.NewContext()
+	ctx.Store = s.store
+	times := 0
```
Comment (mjonss) [inline ddl/db_test.go:7793]:
```suggestion
```

File: ddl/db_test.go [hunk 4]
```diff
@@ -7772,3 +7772,57 @@ func (s *testSerialDBSuite) TestAddGeneratedColumnAndInsert(c *C) {
...
+	tk.MustExec("use Issue32079")
+	tk.MustExec("create table t (id bigint unsigned not null primary key, c1 int, key(c1))")
+	tk.MustExec("insert into t values (20, 20), (40, 40), (80, 80), (160, 160);")
+
+	var checkErr error
+	tk1 := testkit.NewTestKit(c, s.store)
+	_, checkErr = tk1.Exec("use Issue32079")
+
+	d := s.dom.DDL()
+	hook := &ddl.TestDDLCallback{Do: s.dom}
```
Comment (mjonss) [inline ddl/db_test.go:7789]:
```suggestion
	originHook := d.GetHook()
	defer d.(ddl.DDLForTest).SetHook(originHook)
	hook := &ddl.TestDDLCallback{Do: s.dom}
```

File: ddl/db_test.go [hunk 5]
```diff
@@ -7772,3 +7772,57 @@ func (s *testSerialDBSuite) TestAddGeneratedColumnAndInsert(c *C) {
...
+	var currJob *model.Job
```
Comment (mjonss) [inline ddl/db_test.go:7790]:
```suggestion
```

File: ddl/db_test.go [hunk 6]
```diff
@@ -7772,3 +7772,57 @@ func (s *testSerialDBSuite) TestAddGeneratedColumnAndInsert(c *C) {
...
+		case model.StateWriteReorganization:
+			if job.SchemaState == model.StateWriteReorganization && times == 0 {
+				_, checkErr = tk1.Exec("update t set id = 16 where id=160;")
+				if checkErr != nil {
+					return
+				}
+				currJob = job
+				times++
+			}
+			//*/
```
Comment (mjonss) [inline ddl/db_test.go:7819]:
```suggestion
```

**Review by mjonss — APPROVED** [review]
LGTM, just needs some minor cleanup

File: ddl/db_test.go [hunk 7]
```diff
@@ -7770,3 +7770,49 @@ func (s *testSerialDBSuite) TestAddGeneratedColumnAndInsert(c *C) {
 	tk.MustQuery("select * from t1 order by a").Check(testkit.Rows("4 5", "10 11"))
 	c.Assert(checkErr, IsNil)
 }
+
+func (s *testSerialDBSuite) TestIssue32079(c *C) {
```
Comment (zimulala) [inline ddl/db_test.go:7774]: Could we use `TestDeleteOnlyForDropColumnWithIndexes` directly? We can add `update t set id = 4 where id=40;` to `sqls` in `TestDeleteOnlyForDropColumnWithIndexes` .

**Comment by zimulala** [comment]
Please fill out `What is changed and how it works?`.

**Review by zimulala — COMMENTED** [review]
Do we need to fix this issue on `drop columns` operation?

**Review by zimulala — APPROVED** [review]
LGTM

**Comment by zimulala** [comment]
@wjhuang2016 Please fix CI

**Pull request merged**